### PR TITLE
ROL Trilinos Sync 09-Nov-2024

### PR DESCRIPTION
**Pull request by dridzal**
@trilinos/rol 

## Description
Regular merge between the ROL-Trilinos repo and the main Trilinos repo, via a Trilinos fork.

## Motivation and Context
Bug fixes and new features.

## How Has This Been Tested?
RedHat 9 Container testing passes locally, using gcc-10.3.0, performed in ROL-Trilinos (no forward dependencies) and in the Trilinos fork (with forward dependencies).

### CODE REVIEW

**Comment by marcfehling** [comment]
Any chance getting #13560 on board?

**Comment by dridzal** [comment]
> Any chance getting #13560 on board?

@marcfehling , we are trying to perform a standard two-way merge between ROL and Trilinos with a somewhat limited set of changes.  I believe that @gregvw has already addressed #13560 locally, but he hasn't pushed the changes.  They would be included in the next merge, which we would perform shortly after this one is completed.

At the moment, it appears that ROL commits are missing git signoffs, however the missing signoffs are _**before**_ the agreed-upon enforcement date.  In other words, the logic to check the signoffs is broken. @trilinos/framework , please fix the logic to only verify signoffs before the agreed-upon date, thx.   I notified ROL developers of the signoff requirement as soon as I was notified by @trilinos/framework , and they have obliged.

CC: @dpkouri @aj463 @sebrowne @bartlettroscoe

**Comment by mperego** [comment]
@dridzal, the fastest solution might be to squash the commits in this PR (assuming you are OK losing the history of the changes) in one commit, sign it off and force push. 
@trilinos/framework team might be willing to merge this as is.

**Comment by dridzal** [comment]
> @dridzal, the fastest solution might be to squash the commits in this PR (assuming you are OK losing the history of the changes) in one commit, sign it off and force push. @trilinos/framework team might be willing to merge this as is.

@mperego :  I agree that it may be the fastest solution, but we would prefer a cleaner approach.  I don't think that it is fair to require commits from early 2024 be subject to these requirements.  Where would it end?  All of Trilinos, starting with the very first commit?  I think there should be a cutoff date, as I was told in the Trilinos Developers' OPEN Meeting on August 6, 2024.  At that point, the date was not determined yet, and I didn't think the requirement would be retroactive, without a clear cutoff date.

**Comment by bartlettroscoe** [comment]
> > @dridzal, the fastest solution might be to squash the commits in this PR (assuming you are OK losing the history of the changes) in one commit, sign it off and force push. @trilinos/framework team might be willing to merge this as is.
> 
> @mperego : I agree that it may be the fastest solution, but we would prefer a cleaner approach. I don't think that it is fair to require commits from early 2024 be subject to these requirements. Where would it end? All of Trilinos, starting with the very first commit? I think there should be a cutoff date, as I was told in the Trilinos Developers' OPEN Meeting on August 6, 2024. At that point, the date was not determined yet, and I didn't think the requirement would be retroactive, without a clear cutoff date.

Putting requirements on individual Git commits like this is asking for frustrating situations like this.  It was unavoidable given the nature of distributed development with Git.  (I predict we will be hearing about cases like this for many months to come.)  But putting requirements on individual commits like this is reasonable moving forward.

My opinion on how things should move forward:

(1) Update the Trilinos GHA check for the signoff to look at the author date and only fail the commit if it is after a certain date.  (That should fix the ROL problem?).   If that can't be done on the Trilinos side, just for this sync, quash the commits from the ROL side and sync to Trilinos repo 'develop'.  Then when you merge from Trilinos back to ROL, you can resolve the conflicts to make things go more smoothly in the future.   That will create slightly messy Git history, but it should go pretty smoothly.

(2) Encourage all Trilinos and ROL developes to install a local git push hook to stop pushing commits that are missing the signoff line.  I wrote a such a hook that I am now using in all of my local git repo.  See [pre-push](https://github.com/TriBITSPub/TriBITS/blob/master/common_tools/git/hooks/client-side/pre-push). (That is a lot of code for just checking for the signoff line, but it would be easy to extend to allow arbitrary pre-push commit checks.)  This hook can be installed in any Git repo by cloning the TriBITS repo and running:

```
$ cd Trilinos/

$ <TriBITS-base-dir>/common_tools/git/hooks/client-side/copy_hooks_scripts.sh
```

(The `commit-msg` check runs codespell on the commit msg if codespell is in path.)

Those could be made easier to install directly from the Trilinos repo if there was interest.

@dridzal, let me know if you want help with any of that.

**Comment by bartlettroscoe** [comment]
> Encourage all Trilinos and ROL developes to install a local git push hook to stop pushing commits that are missing the signoff line.

@dridzal, you should also consider installing and using the same [DCO GHA](https://github.com/trilinos/Trilinos/blob/develop/.github/workflows/per-commit.yml) in the ROL GitHub repo as a post-merge check (that sends out notifications).  That will let your developers know there is problem early.  (But fixing it will be hard unless you allow force pushes to 'master'.)  But recommendation would be for developers to run these checks pre-push to avoid the problems once they show up in your main GitHub repo 'master' branch.

**Comment by ccober6** [comment]
Sorry that you are running into this @ridzal.  We announced that DCO was coming six months ago, but only recently started enforcing it.  Unfortunately, I do not think @bartlettroscoe suggestion of only failing if the author and date is after a certain date will work.  I believe we have the problem that no matter when the code was committed to remote repos, the code is only contributed to Trilinos when it is merged to the GitHub repo and then requires the DCO check.  So whether the code was committed yesterday or 20 years ago, we need to ensure that the authors complete the DCO for the merge to Trilinos.

**Comment by dridzal** [comment]
@bartlettroscoe :  Thanks for the pointers.  We definitely intend of preventing any pushes without signoffs into the ROL repo.  But to clarify, ROL developers have been using signoffs for several months now.

@ccober6 : Can you remind me, why are we doing the signoffs?

@ccober6 , @bartlettroscoe :  We need a solution that does not involve force push.  Please advise.

**Comment by ccober6** [comment]
This is a Linux Foundation requirement that all contributors state that they agree to the terms published at https://developercertificate.org/ for that particular contribution (e.g., contributors are authors and they have the right to submit it under the open source license.  See link for full details).

**Comment by bartlettroscoe** [comment]
> But to clarify, ROL developers have been using signoffs for several months now.

I can verify that the most recent ROL commit that does not contain the 'signoff' line was 8/19/2024 by running:

```
$ git log-short --graph --grep=Signed-off-by --invert-grep --no-merges rol-trilinos/develop  --not github/develop
```

which shows:

```
* babef0cdd20 "Fix shadowing warning"
  Author: Aurya Javeed <asjavee@sandia.gov>
  Date:   Mon Aug 19 10:30:38 2024 -0600 (3 months ago)

...
```

Note that Trilinos 'develop' has many **non-merge** commits that don't contain the signoff line as recent as 10/17/2024 (i.e. 4 weeks ago) as shown by:

```
$ git log-short --graph --grep=Signed-off-by --invert-grep --no-merges github/develop

* ffcdd924ff6 "MueLu: test higher-order line, quad in IntrepidPCoarsenFactory.cpp (#13534)"
  Author: Nate Roberts <nvrober@sandia.gov>
  Date:   Thu Oct 17 11:20:08 2024 -0500 (4 weeks ago)
  
* 8a019ae517e "Fix typo"
  Author: Samuel Browne <sebrown@sandia.gov>
  Date:   Tue Oct 15 14:07:19 2024 -0600 (4 weeks ago)
  
* 6a817dfbf27 "Intrepid2: increase max orders for basis, cubature (#13505)"
  Author: Nate Roberts <nvrober@sandia.gov>
  Date:   Thu Oct 10 14:40:18 2024 -0500 (5 weeks ago)
  
* f88430eee3c "belos: Fix shadow warning reported in #13512."
  Author: Victor Brunini <vebruni@sandia.gov>
  Date:   Tue Oct 8 11:32:04 2024 -0600 (5 weeks ago)
  
* 4bfd20d24d7 "Trilinos: Update README.md for OpenSSF Best Practices Badge (#13513)"
  Author: Chris Siefert <csiefer2@users.noreply.github.com>
  Date:   Tue Oct 8 14:20:06 2024 -0600 (5 weeks ago)

...
```

How does this work?

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe : We need a solution that does not involve force push. Please advise.

@dridzal, if ROL is not willing to rebase and force push the commits from the last 3 months, then only some type of snapshotting process would work.  There are two options that I can think of:

1.	Squash and sync all changes on the ROL repo not pushed to Trilinos and maintain a sperate sync integration branch in the ROL repo to manage future syncs back and forth to and from Trilinos.  (This is basically a form of snapshotting but just involved as single shared git history.)

2.	Move ROL into its own Git repo and switch ROL to being snapshotted into Trilinos (and make ROL developers clone the ROL repo under Trilinos and configure with `-DROL_SORUCE_DIR_OVERRIDE=rol` and use gitdist to pull and push).

Option 1 will be easier short-term but will complicate syncs between the Trilinos and ROL repos/branches and complicate ROL and Trilinos Git history.  But ROL developers can just pull and push to a single Git repo like before.  It is just that the sync process to and from Trilinos will get more complex.

Option 2 gives you the most flexibility and makes sure that Trilinos policies don’t impact ROL workflows and visa-vera.  But it creates extra complexity for ROL developers in addition to the person who does the sync process for ROL and Trilinos.

Let me know if you want to talk through these and draw some pictures.  (I have my SNL iPad so I can draw good virtual diagrams with MS Teams whiteboard now 😊)

I don’t see other solutions yet off the top of my head that ROL can control.  Does anyone else have any ideas?

---- 

P.S. I would not discount rebasing and force pushing.  It involves some short-term pain (and force-pushing some ROL branches) but keeps the future workflows simple (other than carefully checking that all commits contain the signoff line).  But after this initial work, ROL developers and syncs to and from Trilinos can keep going on just they have been for years.

So either ROL or Trilinos will need to bend a policy in the short-term, or ROL developers are going to have a more complex workflow going forward.

**Comment by sebrowne** [comment]
@dridzal I have spoken with @ccober6 are willing to bend policy to allow this PR through.  But no future PRs.  ROL needs to decide where its development history can/should live.  If it is going to live in the Trilinos repository (as 'a part of Trilinos'), it needs to comply with the same policy (and it sounds like you will be doing this going forwards).  If not, then a snapshot approach is one simple way to handle the integration.

What is the driver for having a separate ROL repository, but then keeping all of the development history when it is added to the Trilinos repository?

What @ccober6 stated earlier was correct.  The requirement applies to changes to Trilinos from this point forwards.  Commit date is irrelevant.  Any change that gets submitted through a PR now will have to be comprised entirely of commits that have been signed-off, as those commits are "new" as far as Trilinos' `develop` history is concerned.

**Comment by bartlettroscoe** [comment]
> What is the driver for having a separate ROL repository, but then keeping all of the development history when it is added to the Trilinos repository?

@sebrowne, see "Addition of a subteam branch" on slide 9 of:

* https://bartlettroscoe.github.io/publications/DesignPatternsForGitWorkflows.pdf

And it is not just ROL that uses this type of workflow.

This is just one simple type of a truly distributed Git workflow.

**Comment by sebrowne** [comment]
If review, etc. is happening in another repository, then the history should probably be interacted with in that repository as well.

If the history is desired in Trilinos, then all commits to that other repository will need to comply with policies set in the "main" project repository (Trilinos), and keeping all of those checks/etc. synchronized in that subteam repository (ROL) is a maintenance cost of having that subteam repository (else issues will not be caught until integration time, as is the case here, which may necessitate rebasing the inbound commits).

**Comment by sebrowne** [comment]
@dridzal let me know when this is ready to merge and I will allow it in past the DCO signoff requirement.  Looks like it needs a code review, but other than that the required checks have passed.

File: .github/workflows/update_rol_parameters.yml
Comment (cgcgcg) [inline .github/workflows/update_rol_parameters.yml:1]: Are you sure this should be pushed? It looks like the `find_parameters.sh` script is missing.

**Review by cgcgcg — CHANGES_REQUESTED** [review]

File: .github/workflows/update_rol_parameters.yml
Comment (dridzal) [reply .github/workflows/update_rol_parameters.yml:1]: @gregvw , please advise.

Comment (gregvw) [reply .github/workflows/update_rol_parameters.yml:1]: @cgcgcg I am confused about your statement that  "`find_parameters.sh` script is missing." It is [here](https://github.com/Rapid-Optimization-Library/ROL-Trilinos/blob/develop/packages/rol/rol_parameters/find_parameters.sh)

**Comment by dridzal** [comment]
> @dridzal let me know when this is ready to merge and I will allow it in past the DCO signoff requirement. Looks like it needs a code review, but other than that the required checks have passed.
> 
> Ideally, for this one, can you state the signoff text in a comment as part of this PR? It's not perfect, but it's better than not having it at all.

@sebrowne , I'll let you know when we are ready.  At that point, I'll also state in a comment the signoff text.  Thank you for accommodating this exception!

To summarize, the main issue here is not that ROL has not been complying with the signoff requirement, it is that there was some residual history, which we could not merge earlier.  When a cutoff date was discussed, I must have misunderstood it as something that the DCO scripts would check.  The Framework team meant a different type of cutoff.

Going forward, we would prefer to stay fully integrated with Trilinos, in the spirit of a fully distributed workflow.  We have adopted the signoff requirements already, as evidenced by the Aug. 19 time stamp of the latest non-signed commit.  We will also automatically enforce the signoff requirement in the ROL repo.

File: .github/workflows/update_rol_parameters.yml
Comment (achauphan) [inline .github/workflows/update_rol_parameters.yml:38]: @ccober6 @sebrowne should the commits made by this action require a sign-off by a ROL member?

File: .github/workflows/update_rol_parameters.yml
Comment (cgcgcg) [inline .github/workflows/update_rol_parameters.yml:22]: @gregvw I think this step will run the root of the Trilinos checkout, not in `packages/rol/rol_parameters`.

File: .github/workflows/update_rol_parameters.yml
Comment (cgcgcg) [inline .github/workflows/update_rol_parameters.yml:25]: Same here.

**Review by gregvw — DISMISSED** [review]
We should remove remove the file .github/workflows/update_rol_parameters.yml after the merge

File: .github/workflows/update_rol_parameters.yml
Comment (gregvw) [reply .github/workflows/update_rol_parameters.yml:22]: Ah yes. This file should be removed.

File: .github/workflows/update_rol_parameters.yml
Comment (dridzal) [reply .github/workflows/update_rol_parameters.yml]: Removed file.

File: .github/workflows/update_rol_parameters.yml
Comment (dridzal) [reply .github/workflows/update_rol_parameters.yml]: Removed file.

File: .github/workflows/update_rol_parameters.yml
Comment (dridzal) [reply .github/workflows/update_rol_parameters.yml]: Removed file.

**Review by aj463 — APPROVED** [review]

**Comment by dridzal** [comment]
@sebrowne :  After one minor modification (removed an unnecessary file), we are ready to merge.  The testing has started, and I hope all tests will pass as before.

Signed-off-by: Denis Ridzal <dridzal@sandia.gov>

**Pull request merged**
